### ISSUE TITLE
Suppress metadata changes at terraform plan?

### ISSUE THREAD

**Issue by mattpopa** [hashicorp/terraform-provider-helm#1315]
Older helm provider versions didn't show this `metadata` changes when terraform detected a change/run plan, and the new helm provider versions are posting a wall of text of `metadata` changes which have no real value (to me) and just clog up my tf plan output.

example doing a tf plan on a helm resource where we only updated the image tag var:

```
  # module.api-eks.helm_release.app will be updated in-place
  ~ resource "helm_release" "app" {
        id                         = "api"
      ~ metadata                   = [
          - {
              - app_version = "latest"
              - chart       = "api"
              - name        = "api"
              - namespace   = "production"
              - revision    = 1420
              - values      = jsonencode(
                    {
                      - appConfig          = {
                          - apiBackendReplicationTaskId       = "none"
                          - applicationMode                   = "none"
                          - baseApiUrl                        = "none"
                          - something = "else"
                          - foo = "bar"
                          - 
                          < it goes on for many >
                          < many >
                          < lines >
                          < and it's of no value >
                          < just noise on tf plan >
                          
```

let's say we only have a field  `deploymentTimestamp` updated.  We'd rather see the changed field only on terraform plan, and suppress the whole `metadata` update, e.g. terraform plan should only show

```
  # module.api-eks.helm_release.app will be updated in-place
  ~ resource "helm_release" "app" {
        id                         = "api"
        [...]
        # (25 unchanged attributes hidden)

      + set {
          + name  = "deploymentTimestamp"
          + value = "19012024-225905"
        }

        # (62 unchanged blocks hidden)
        
```
This way the terraform plan is clear and concise, more human (easier to read/follow) without the metadata removal. Does it make sense?

## Terraform version, Kubernetes provider version and Kubernetes version

```
Terraform version: v1.6.5
Helm Provider version: v2.12.0 (same on v2.12.1)
Kubernetes version: v2.24.0
```
## Terraform configuration

```hcl
resource "helm_release" "app" {
  namespace       = var.namespace != "" ? var.namespace : terraform.workspace
  chart           = var.chart_name
  version         = var.chart_version
  name            = var.app_name
  timeout         = var.deployment_timeout
  cleanup_on_fail = var.helm_cleanup_on_fail
  atomic          = var.helm_atomic_creation
  max_history     = var.helm_max_history
  wait            = var.helm_wait_for_completion

  dynamic "set" {
    for_each = local.k8s_app

    content {
      name  = set.key
      value =  set.value
    }
  }

  values = var.some_ingress_values
}

```

## Question

Is there any way to suppress the `metadata` changes at terraform plan?

**Comment by MiguelNdeCarvalho** [comment]
I'm really curious if this is possible to do, as I'm using the https://registry.terraform.io/providers/hashicorp/helm/latest/docs#manifest experiment.

**Comment by eklesel** [comment]
Would also be very interested in this, we have a fairly hefty internal helm chart with a few dependencies and a large values input spread across multiple files. When a single values file is updated (even just adding a space), the entire metadata is recomputed and showing as changed, although the value is not known as it's computed.

From a recent plan, the metadata takes up 2183 lines of a total 2392, we're currently forced to redirect stdout of apply/plan runs to a text to then read the plan, and in CI it causes huge job logs which is completely unnecessary.

For us, the metadata of the helm chart is useless and we'd like to disable the computation if possible (or at least hide it from output).

We already mark the values file as sensitive because the output is similarly huge, it outputs the entire text file before/after instead of just showing the values which have changed/added/removed. For large values files this is useless for comparison, so we compare changes using the manifest experiment instead which works very well.

**Comment by sbeginCoveo** [comment]
We're in the same situation, we managed to silence the values diff, but now were back at square one with a huge and useless diff in our terraform plans
```
      ~ values                     = (sensitive value)
```

**Comment by eklesel** [comment]
Another point which is probably of more serious concern, the metatadata output will show sensitive values.

When changing **any** values in the helm_release resource the **entire** metadata output is shown, even if sensitive values aren't changed.

**Comment by BenB196** [comment]
> Another point which is probably of more serious concern, the metatadata output will show sensitive values.
> 
> When changing **any** values in the helm_release resource the **entire** metadata output is shown, even if sensitive values aren't changed.

I don't believe this is (at least 100%) accurate, at least on the latest version `2.12.1`.

![image](https://github.com/hashicorp/terraform-provider-helm/assets/8277432/e3d1b4e8-244a-46c5-8ec7-ff2163cba6b7)

Shows that some values that were set using `set_sensitive` are being properly hidden

**Comment by eklesel** [comment]
Apologies, @BenB196 is correct, I was mistaken as we do not use `set_sensitive` to set sensitive variables, therefore they're displayed.

It would be useful to have a `values_sensitive` parameter so that we can set sensitive values using freeform YAML, but that's another issue.

**Comment by jsantos17** [comment]
Interestingly, I find 

```
  lifecycle {
    ignore_changes = [
      metadata
    ]
  }
```

successfully suppressed the metadata junk. However, the following (erroneous) warning ensues:

```
│ Adding an attribute name to ignore_changes tells Terraform to ignore future changes to the argument in
│ configuration after the object has been created, retaining the value originally configured.
│
│ The attribute metadata is decided by the provider alone and therefore there can be no configured value to compare
│ with. Including this attribute in ignore_changes has no effect. Remove the attribute from ignore_changes to quiet
│ this warning.
```

This does have an effect, despite the warning.

**Comment by imranismail** [comment]
Hello folks, if you run into this issue and want a temporary fix, you can use my script detailed here: https://gist.github.com/imranismail/3364db5b49200247ca8cb675c96589fe

**Comment by multipletwigs** [comment]
out of curiosity, why did this metadata field existed in the first place? does it help with any sort of identification for the resource?

**Comment by pragnarok** [comment]
No updates yet? I'm going through the same situation and it's really annoying.

**Comment by PPACI** [comment]
Update: This commit seems to be linked to this issue https://github.com/hashicorp/terraform-provider-helm/commit/62d34d271933c35561118e925aeced1bee077987

**Comment by ghost** [comment]
Did a check with the latest release - [v2.16.0](https://github.com/hashicorp/terraform-provider-helm/releases/tag/v2.16.0)

### After
<img width="724" alt="Screenshot 2024-10-11 at 16 42 03" src="https://github.com/user-attachments/assets/a50eed7f-b4fa-42b6-aed4-37313caede49">


### Before
<img width="724" alt="Screenshot 2024-10-11 at 16 33 24" src="https://github.com/user-attachments/assets/7ee0d17a-25ae-4073-8852-d35535582d31">

**Comment by ghost** [comment]
Hm, but it works differently in other cases?

It still show a lot of irrelevant output when values are changed.

**Comment by machisuji** [comment]
> Did a check with the latest release - [v2.16.0](https://github.com/hashicorp/terraform-provider-helm/releases/tag/v2.16.0)
> 
> ### After
> <img alt="Screenshot 2024-10-11 at 16 42 03" width="724" src="https://private-user-images.githubusercontent.com/88528265/375775374-a50eed7f-b4fa-42b6-aed4-37313caede49.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEwNTg2NzMsIm5iZiI6MTczMTA1ODM3MywicGF0aCI6Ii84ODUyODI2NS8zNzU3NzUzNzQtYTUwZWVkN2YtYjRmYS00MmI2LWFlZDQtMzczMTNjYWVkZTQ5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTA4VDA5MzI1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc3ZGE2NGZiMmQwZjMyM2RmNTFhMTY1MjQyNWUyZmEwMjJlZjZjOGM0MWU1NDM3N2EzNzFjOTgxNGYyNjk5ZmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.GJ7-LjjcBj2JEy1_W30FhM4O7DBSEInSfDkORgeHdJQ">
> ### Before
> <img alt="Screenshot 2024-10-11 at 16 33 24" width="724" src="https://private-user-images.githubusercontent.com/88528265/375774040-7ee0d17a-25ae-4073-8852-d35535582d31.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEwNTg2NzMsIm5iZiI6MTczMTA1ODM3MywicGF0aCI6Ii84ODUyODI2NS8zNzU3NzQwNDAtN2VlMGQxN2EtMjVhZS00MDczLTg4NTItZDM1NTM1NTgyZDMxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTA4VDA5MzI1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA5YTcxZGQzMTU1MzYyN2FlZGVmMWMyOTQ4ZGU0ZWM1NWVjYTJkZTZiYzM4MzM2MmI3OGY2NTc0ZmQwODZjYzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.MS7Xv9CAbvIoMn20Es8lzuApuvs-9f8Hwv97lNzZlII">

I've just upgraded to [v2.16.0](https://github.com/hashicorp/terraform-provider-helm/releases/tag/v2.16.0) as well, but it has changed nothing. I still get all the metadata in the output with each apply like in your before, sadly.

**Comment by machisuji** [comment]
What does help is indeed to add the 

```
lifecycle {
  ignore_changes = [ metadata ]
}
```

as mentioned further up.

The warning is a bit annoying though. It seems to give the correct hint for the issue at the same time:

> The attribute metadata is decided by the provider alone and therefore there can be no configured value to compare with.

Since there is "no configured value to compare with" it always shows a diff looking as if everything has been removed, I suppose.

I might just accept that warning and live with it because it's better than outputting all the (partly sensitive) metadata with each apply.

**Comment by SoraMakes** [comment]
for me this
>   lifecycle {
    ignore_changes = [metadata[0]]
  }

removes the metadata changes from the plan step and does not produce the warning.

**Comment by guidodobboletta** [comment]
Any updates on this? This not only produces a lot of output that doesn't add real value but it also prints the secrets inside the actual diff the plans which is big security consideration

```
  # helm_release.argocd will be updated in-place
  ~ resource "helm_release" "argocd" {
        id                         = "argocd"
      ~ metadata                   = [
          - {
              - app_version    = "v2.14.11"
              - chart          = "argo-cd"
              - first_deployed = 1745952790
              - last_deployed  = 1746134744
              - name           = "argocd"
              - namespace      = "argocd"
              - revision       = 19
              - values         = jsonencode(
                    {
                      - configs        = {
                          - cm     = {
                              - "admin.enabled"               = false
                              - "dex.config"                  = <<-EOT
                                    connectors:
                                    - config:
                                        issuer: https://accounts.google.com
                                        clientID: redacted
                                        clientSecret: redacted
                                      type: oidc
                                      id: google
                                      name: Google
                  ..........

      ~ values = [
          - (sensitive value),
          + (sensitive value),
        ]
```

**Comment by eklesel** [comment]
> What does help is indeed to add the
> 
> ```
> lifecycle {
>   ignore_changes = [ metadata ]
> }
> ```
> 
> as mentioned further up.
> 
> The warning is a bit annoying though. It seems to give the correct hint for the issue at the same time:
> 
> > The attribute metadata is decided by the provider alone and therefore there can be no configured value to compare with.
> 
> Since there is "no configured value to compare with" it always shows a diff looking as if everything has been removed, I suppose.
> 
> I might just accept that warning and live with it because it's better than outputting all the (partly sensitive) metadata with each apply.

This workaround seems to have broken with version 3 (I'm running 3.0.2). The structure of the metadata parameter has changed from a list of maps/json objects to the following

```
      ~ metadata                   = {
          ~ app_version    = <...> -> (known after apply)
          ~ chart          = <...> -> (known after apply)
          ~ first_deployed = <...>-> (known after apply)
          ~ last_deployed  = <...>-> (known after apply)
          ~ name           = <...> -> (known after apply)
          ~ namespace      = <...> -> (known after apply)
          ~ notes          = <<-EOT
            <large multistring>
            EOT -> (known after apply)
          ~ revision       = 19 -> (known after apply)
          ~ values         = jsonencode(
              <huge HCL>
            ) -> (known after apply)
          ~ version        = <...> -> (known after apply)

```

I tried `metadata.app_version`, `metadata["app_version"]`, `metadata[0]` as well as just `metadata` but it still seems to show in the plan. Now that the values are expanded into HCL, it seems even more verbose than before given that they're still exported as their own parameter.

Anyone else having the same issue?

**Comment by maschsyn** [comment]
Yes, I also noticed this workaround recently broke for me, although I didn't dig into exactly why.

**Comment by deadlysyn** [comment]
with latest provider i can still use ignore_changes = [metadata] to get a clean plan, but on apply it leads to a new issue:

```
  │ Error: Provider produced inconsistent result after apply
  │
  │ When applying changes to helm_release.aws_load_balancer_controller,
  │ provider "provider[\"registry.terraform.io/hashicorp/helm\"]" produced an
  │ unexpected new value: .metadata.last_deployed: was
  │ cty.NumberIntVal(1.756338773e+09), but now
  │ cty.NumberIntVal(1.778598409e+09).
  │
  │ This is a bug in the provider, which should be reported in the provider's
  │ own issue tracker.
  ╵
  ╷
  │ Error: Provider produced inconsistent result after apply
  │
  │ When applying changes to helm_release.aws_load_balancer_controller,
  │ provider "provider[\"registry.terraform.io/hashicorp/helm\"]" produced an
  │ unexpected new value: .metadata.revision: was cty.NumberIntVal(2), but now
  │ cty.NumberIntVal(3).
  │
  │ This is a bug in the provider, which should be reported in the provider's
  │ own issue tracker.
```
